File: Medium_Python/144_Binary_tree_preorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res=[]
        def preorderTraversalHelper(root):
            if root:
                res.append(root.val)
                preorderTraversalHelper(root.left)
                preorderTraversalHelper(root.right)

        preorderTraversalHelper(root)
        return res

File: Medium/144_Binary_tree_preorder_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        stack<TreeNode*> rootstack;
        vector<int> res;
        if(root!=nullptr)
        {
            rootstack.push(root);
            while(!rootstack.empty())
            {
                TreeNode* cur = rootstack.top();
                rootstack.pop();
                res.push_back(cur->val);
                if(cur->right!=nullptr)
                rootstack.push(cur->right);
                if(cur->left!=nullptr)
                rootstack.push(cur->left);
            }
        }
        return res;
    }
};
